application/src/main/java/org/togetherjava/tjbot/commands/moderation/BanCommand.java

```diff
@@ -0,0 +1,101 @@
+
+import java.util.Objects;
+
+/**
+ * This command can ban users and optionally remove their messages from the past days. Banning can
+ * also be paired with a ban reason. The command will also try to DM the user to inform them about
+ * the action and the reason.
+ * <p>
+ * The command fails if the user triggering it is lacking permissions to either ban other users or
+ * to ban the specific given user (for example a moderator attempting to ban an admin).
+ */
+public final class BanCommand extends SlashCommandAdapter {
+    private static final Logger logger = LoggerFactory.getLogger(BanCommand.class);
+    private static final String TARGET_OPTION = "user";
+    private static final String DELETE_HISTORY_OPTION = "delete-history";
+    private static final String REASON_OPTION = "reason";
+
+    /**
+     * Constructs an instance.
+     */
+    public BanCommand() {
+        super("ban", "Bans the given user from the server", SlashCommandVisibility.GUILD);
+
+        getData().addOption(OptionType.USER, TARGET_OPTION, "The user who you want to ban", true)
+            .addOption(OptionType.STRING, REASON_OPTION, "Why the user should be banned", true)
+            .addOptions(new OptionData(OptionType.INTEGER, DELETE_HISTORY_OPTION,
+                    "the amount of days of the message history to delete, none means no messages are deleted.",
+                    true).addChoice("none", 0).addChoice("recent", 1).addChoice("all", 7));
+    }
+
+    @SuppressWarnings("MethodWithTooManyParameters")
+    private static void banUser(@NotNull User target, @NotNull Member author,
+            @NotNull String reason, int deleteHistoryDays, @NotNull Guild guild,
+            @NotNull SlashCommandEvent event) {
+        event.getJDA()
+            .openPrivateChannelById(target.getIdLong())
+            .flatMap(channel -> channel.sendMessage(
+                    """
+                            Hey there, sorry to tell you but unfortunately you have been banned from the server %s.
+                            If you think this was a mistake, please contact a moderator or admin of the server.
+                            The reason for the ban is: %s
+                            """
+                        .formatted(guild.getName(), reason)))
+            .mapToResult()
+            .flatMap(result -> guild.ban(target, deleteHistoryDays, reason))
+            .flatMap(result -> event.reply("'%s' was banned by '%s' for: %s"
+                .formatted(target.getAsTag(), author.getUser().getAsTag(), reason)))
+            .queue();
+
+        logger.info(
+                "'{}' ({}) banned the user '{}' ({}) from guild '{}' and deleted their message history of the last {} days, for reason '{}'",
+                author.getUser().getAsTag(), author.getId(), target.getAsTag(), target.getId(),
+                guild.getName(), deleteHistoryDays, reason);
+    }
+
+    @Override
+    public void onSlashCommand(@NotNull SlashCommandEvent event) {
```

Comment:
The method should handle the case that an user is already banned better. Right now, it just bans them again. Instead, it should check it, return and send a special message to the user, telling them that the user is already banned.

Reply:
implemented